File: 1063-best-sightseeing-pair/1063-best-sightseeing-pair.cs
public class Solution {
    public int MaxScoreSightseeingPair(int[] values) {
        int n = values.Length;
        List<int> maxSum = new List<int>();
        
        for(int i = 0 ; i < n - 1; i++){
            maxSum.Add(i + values[i]);
        }
        int ans = int.MinValue, mn = int.MaxValue;

        for(int i = n - 2; i >= 0 ;i--){
            mn = Math.Min(mn,(i+1) - values[i+1]);
            ans = Math.Max(ans,maxSum[i] - mn);
        }
        return ans;
    }
}

File: 1063-best-sightseeing-pair/1063-best-sightseeing-pair.cpp
class Solution {
public:
    int maxScoreSightseeingPair(vector<int>& values) {
        int n = values.size();
        vector<int>maxSum(n-1);
        vector<int>minSum(n-1);
        for(int i = 0 ;i<n-1;i++){
            maxSum[i] = i + values[i];
        }
        int mn = INT_MAX;
        for(int i= n-2;i>=0;i--){
            mn = min(mn,(i+1)-values[i+1]);
            minSum[i] = mn;
        }
        int mx = INT_MIN;
        for(int i = 0 ;i<n-1;i++){
            mx = max(mx,maxSum[i] - minSum[i]);
        }
        return mx;
    }
};